hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/upgrade/ReconLayoutVersionManager.java

```diff
@@ -0,0 +1,146 @@
+
+  // Metadata Layout Version (MLV) of the Recon Metadata on disk
+  private int currentMLV;
+  // Software Layout Version (SLV) of the Recon service
+  private final int currentSLV;
+
+  public ReconLayoutVersionManager(ReconSchemaVersionTableManager schemaVersionTableManager) {
+    this.schemaVersionTableManager = schemaVersionTableManager;
+    this.currentMLV = determineMLV();
+    this.currentSLV = determineSLV();
+    ReconLayoutFeature.registerUpgradeActions();  // Register actions via annotation
+  }
+
+  /**
+   * Determines the current Metadata Layout Version (MLV) from the version table.
+   * @return The current Metadata Layout Version (MLV).
+   */
+  private int determineMLV() {
+    return schemaVersionTableManager.getCurrentSchemaVersion();
+  }
+
+  /**
+   * Determines the Software Layout Version (SLV) based on the latest feature version.
+   * @return The Software Layout Version (SLV).
+   */
+  private int determineSLV() {
+    return Arrays.stream(ReconLayoutFeature.values())
+        .mapToInt(ReconLayoutFeature::getVersion)
+        .max()
+        .orElse(0); // Default to 0 if no features are defined
+  }
+
+  /**
+   * Finalizes the layout features that need to be upgraded, by executing the upgrade action for each
+   * feature that is registered for finalization.
+   */
+  public void finalizeLayoutFeatures() {
+    // Get features that need finalization, sorted by version
+    List<ReconLayoutFeature> featuresToFinalize = getRegisteredFeatures();
+
+    for (ReconLayoutFeature feature : featuresToFinalize) {
+      try {
+        // Fetch only the AUTO_FINALIZE action for the feature
+        Optional<ReconUpgradeAction> action = feature.getAction(ReconUpgradeAction.UpgradeActionType.AUTO_FINALIZE);
+        if (action.isPresent()) {
+          // Execute the upgrade action & update the schema version in the DB
+          action.get().execute();
+          updateSchemaVersion(feature.getVersion());
+          LOG.info("Feature {} finalized successfully.", feature.getVersion());
+        }
+      } catch (Exception e) {
+        LOG.error("Failed to finalize feature {}: {}", feature.getVersion(), e.getMessage());
+        break;
+      }
+    }
+  }
+
+  /**
+   * Returns a list of ReconLayoutFeature objects that are registered for finalization.
+   */
+  protected List<ReconLayoutFeature> getRegisteredFeatures() {
+    List<ReconLayoutFeature> allFeatures =
+        Arrays.asList(ReconLayoutFeature.values());
+
+    LOG.info("Current MLV: {}. SLV: {}. Checking features for registration...", currentMLV, currentSLV);
+
+    List<ReconLayoutFeature> registeredFeatures = allFeatures.stream()
+        .filter(feature -> {
+          boolean shouldRegister = feature.getVersion() > currentMLV;
+          if (shouldRegister) {
+            LOG.info("Feature {} (version {}) is registered for finalization.",
+                feature.name(), feature.getVersion());
+          } else {
+            LOG.info(
```

Comment:
This log is not requried as every startup, this log will come even its latest meta version.

Reply:
Done!